README.md
# Image-Processing-CLI-in-Rust [![Build Status](https://travis-ci.org/spejss/Image-Processing-CLI-in-Rust.svg?branch=master)](https://travis-ci.org/spejss/Image-Processing-CLI-in-Rust)

CLI for processing images in Rust. Some implementation is custom and for some functionality it uses 3rd party libraries.

This project uses following libraries:

https://github.com/PistonDevelopers/image

Processed images are being stored in the same folder as the source image using the name of the source image with an appended suffix.

Source: hotelroom.jpg

Copy: hotelroomCopy.jpg

Blur: hotelroomBlur.jpg

Generated histograms are also stored this way.

## Implemented functions

* ***Binary threshold*** -o binaryThreshold -v 200
* ***(NAIVE) Histogram for colors (RGB)*** -o histogram
* ***(NAIVE) Histogram for grayscale images*** -o histogramGrayscale
* ***Average color*** -o average
* ***Copy:*** -o copy
* ***Thumbnail:*** -o thumbnail -v 96
* ***blur:*** -o blur -v 4.0
* ***brighten:*** -o brighten -v 10
* ***huerotate:*** -o huerotate -v 10
* ***contrast:*** -o contrast -v 20.0
* ***grayscale***: -o grayscale
* ***invert*** -o invert

## Examples

### Copy image

```bash
ipcli -o copy --image "D:\Image\hotelroom.jpg"
```

### Create a thumbnail

```bash
ipcli -o thumbnail -v 96 --image "D:\Image\hotelroom.jpg"
```

### Generate a histogram of colors

```bash
ipcli -o histogram -i "D:\Image\hotelroom.jpg"
```

![Example histogram of colors](assets/histogram.png)

```text
IPCLI  0.1
Mikolaj Wawrzyniak <mikolaj.wawrzyniak at fh-dortmund.de>
Basic CLI for image processing

USAGE:
    ipcli [OPTIONS] --image <FILE> --operation <operation>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -i, --image <FILE>             Opens specified image file and uses it for transformations.
    -o, --operation <operation>    Specifies operation to be done on the image
    -v, --value <value>            Value for the transformation. To see what values are needed, check the
                                   documentation.
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
* binary threshold
* histogram for rgb
* histogram for grayscale

## 0.1 - 2017-08-16
### Added
* average color
* blur
* brighten
* copy
* contrast
* grayscale
* huerotate
* invert
* thumbnail


src/io.rs
use image::{self, DynamicImage, ImageBuffer, Rgb};
use std::fs::File;

pub fn openImage(i: &str) -> DynamicImage {
    image::open(i).expect("Opening image failed")
}

pub fn saveImage(img: &DynamicImage, i: &str, operation: &str) {
    let outputPath = generateFilename(i, operation);
    let mut out = File::create(outputPath.0).unwrap();
    let format = match outputPath.1.to_lowercase().as_ref() {
        "jpg" | "jpeg" => Some(image::JPEG),
        "png" => Some(image::PNG),
        "gif" => Some(image::GIF),
        "bmp" => Some(image::BMP),
        "ico" => Some(image::ICO),
        _ => None,
    };
    if let Some(format) = format {
        img.save(&mut out, format).expect("Saving image failed");
    } else {
        println!("Unsupported file format");
    }
}

pub fn saveBuffer(buffer: &ImageBuffer<Rgb<u8>, Vec<u8>>, i: &str, operation: &str) {
    let outputPath = generateFilename(i, operation);
    buffer.save(outputPath.0).unwrap();
}

fn generateFilename(i: &str, operation: &str) -> (String, String) {
    let mut outputPath: String = i.chars().take(i.len() - 4).collect();
    let ext: String = i.chars().skip(i.len() - 3).take(3).collect();
    outputPath.push_str(operation);
    outputPath.push_str(".");
    outputPath.push_str(&ext);
    println!("Output path: {}", outputPath);
    (outputPath, ext)
}


src/histogram.rs
use image::{GenericImage, ImageBuffer, Pixel, Rgb};
use io::{openImage, saveBuffer};
use std::cmp;

const WIDTH: u32 = 255;
const HEIGHT: u32 = 200;

/// Generate histogram of grayscale intensity
pub fn grayscale(i: &str) {
    let operation = "HistogramGrayscale";
    let img = openImage(i);

    let grayscale = img.grayscale();

    // create a vector for storing the number of occurrences of each intensity
    let mut occurrences = vec![0; 255];

    // iterate over each pixel in the image and count the occurrences of each intensity
    let (width, height) = grayscale.dimensions();
    for w in 0..width {
        for h in 0..height {
            let pixel = grayscale.get_pixel(w, h);
            let rgb = pixel.to_rgb();
            occurrences[rgb.data[0] as usize] += 1;
        }
    }

    // find highest value of occurrences, so that we can use it as 100% in the histogram
    let maxValue = *occurrences.iter().max().unwrap();

    let mut buffer = createBuffer();
    draw(&mut buffer, &occurrences, maxValue, [0, 0, 0]);

    saveBuffer(&buffer, &i, &operation);
}

/// Generate histogram of RGB values
pub fn rgb(i: &str) {
    let operation = "Histogram";
    let img = openImage(i);

    // create a vector for storing the number of occurrences of each intensity
    let mut occurrencesR = vec![0; 255];
    let mut occurrencesG = vec![0; 255];
    let mut occurrencesB = vec![0; 255];

    // iterate over each pixel in the image and count the occurrences of each intensity
    let (width, height) = img.dimensions();
    for w in 0..width {
        for h in 0..height {
            let pixel = img.get_pixel(w, h);
            let rgb = pixel.to_rgb();
            occurrencesR[rgb.data[0] as usize] += 1;
            occurrencesG[rgb.data[1] as usize] += 1;
            occurrencesB[rgb.data[2] as usize] += 1;
        }
    }

    // find highest value of occurrences, so that we can use it as 100% in the histogram
    let maxValueR = *occurrencesR.iter().max().unwrap();
    let maxValueG = *occurrencesG.iter().max().unwrap();
    let maxValueB = *occurrencesB.iter().max().unwrap();

    let mut buffer = createBuffer();
    draw(&mut buffer, &occurrencesR, maxValueR, [255, 0, 0]);
    draw(&mut buffer, &occurrencesG, maxValueG, [0, 255, 0]);
    draw(&mut buffer, &occurrencesB, maxValueB, [0, 0, 255]);

    saveBuffer(&buffer, &i, &operation);
}

fn createBuffer() -> ImageBuffer<Rgb<u8>, Vec<u8>> {
    let mut buffer = ImageBuffer::<Rgb<u8>, Vec<u8>>::new(WIDTH, HEIGHT);
    for w in 0..WIDTH {
        for h in 0..HEIGHT {
            buffer.get_pixel_mut(w, h).data = [255, 255, 255];
        }
    }
    buffer
}

fn draw(buffer: &mut ImageBuffer<Rgb<u8>, Vec<u8>>, values: &[u32], max: u32, data: [u8; 3]) {
    for i in 0..255_u8 {
        let mut height = cmp::min(
            ((values[usize::from(i)] as f32 / max as f32) * 200.0) as u32,
            u32::from(u8::max_value()),
        ) as u8;
        let c = (HEIGHT - 1).saturating_sub(u32::from(height));
        let mut pixel = buffer.get_pixel_mut(u32::from(i), c);
        pixel.data = data;
    }
}


src/main.rs
#![allow(non_snake_case)]

extern crate clap;
extern crate image;

mod histogram;
mod io;

use clap::{App, Arg};
use image::{FilterType, GenericImage, Pixel, Rgb};
use io::{openImage, saveBuffer, saveImage};

fn main() {
    let matches = App::new("IPCLI")
        .version("0.1")
        .author("Mikolaj Wawrzyniak <mikolaj.wawrzyniak at fh-dortmund.de>")
        .about("Basic CLI for image processing")
        .arg(Arg::with_name("operation")
            .short("o")
            .long("operation")
            .help("Specifies operation to be done on the image")
            .takes_value(true)
            .required(true))
        .arg(Arg::with_name("value")
            .short("v")
            .long("value")
            .help("Value for the transformation. To see what values are needed, check the documentation.")
            .takes_value(true)
            .required(false))
        .arg(Arg::with_name("image")
            .short("i")
            .long("image")
            .value_name("FILE")
            .help("Opens specified image file and uses it for transformations.")
            .takes_value(true)
            .required(true))
        .get_matches();

    let imagePath = matches.value_of("image").unwrap_or("empty");
    let operation = matches.value_of("operation").unwrap_or("empty");
    let value = matches.value_of("value").unwrap_or("empty");

    match operation {
        "average" => averageColor(imagePath),
        "copy" => copy(imagePath),
        "thumbnail" => {
            let size: u32 = value.parse().unwrap();
            createThumbnail(imagePath, size)
        }
        "blur" => {
            let v: f32 = value.parse().unwrap();
            gaussianBlur(imagePath, v)
        }
        "brighten" => {
            let v: i32 = value.parse().unwrap();
            brighten(imagePath, v)
        }
        "huerotate" => {
            let v: i32 = value.parse().unwrap();
            huerotate(imagePath, v)
        }
        "contrast" => {
            let v: f32 = value.parse().unwrap();
            contrast(imagePath, v);
        }
        "grayscale" => {
            grayscale(imagePath);
        }
        "invert" => {
            invert(imagePath);
        }
        "histogramGrayscale" => histogram::grayscale(imagePath),
        "histogram" => histogram::rgb(imagePath),
        "binaryThreshold" => {
            let v: u8 = value.parse().unwrap();
            binary_threshold(imagePath, v);
        }
        _ => println!("Operation not recognised!"),
    }
}

fn averageColor(i: &str) {
    let img = openImage(i);
    let mut r: u32 = 0;
    let mut g: u32 = 0;
    let mut b: u32 = 0;
    let (width, height) = img.dimensions();
    for x in 0..width {
        for y in 0..height {
            let px = img.get_pixel(x, y);
            let rgb = px.to_rgb();
            r = (r + u32::from(rgb.data[0])) / 2;
            g = (g + u32::from(rgb.data[1])) / 2;
            b = (b + u32::from(rgb.data[2])) / 2;
        }
    }
    println!("Average color is: RGB {} {} {}", r, g, b);
}

fn createThumbnail(i: &str, size: u32) {
    let operation = "Thumbnail";
    let img = openImage(i);
    let thumbnail = img.resize(size, size, FilterType::Lanczos3);
    saveImage(&thumbnail, &i, &operation);
}

fn copy(i: &str) {
    let operation = "Copy";
    let img = openImage(i);
    saveImage(&img, &i, &operation);
}

fn gaussianBlur(i: &str, v: f32) {
    let operation = "GuassianBlur";
    let img = openImage(i);
    let blurred = img.blur(v);
    saveImage(&blurred, &i, &operation);
}

fn brighten(i: &str, v: i32) {
    let operation = "Brighten";
    let img = openImage(i);
    let brightened = img.brighten(v);
    saveImage(&brightened, &i, &operation);
}

fn huerotate(i: &str, v: i32) {
    let operation = "Huerotate";
    let img = openImage(i);
    let huerotated = img.huerotate(v);
    saveImage(&huerotated, &i, &operation);
}

fn contrast(i: &str, v: f32) {
    let operation = "AdjustContrast";
    let img = openImage(i);
    let contrast = img.adjust_contrast(v);
    saveImage(&contrast, &i, &operation);
}

fn grayscale(i: &str) {
    let operation = "Grayscale";
    let img = openImage(i);
    let grayscale = img.grayscale();
    saveImage(&grayscale, &i, &operation);
}

fn invert(i: &str) {
    let operation = "Invert";
    let mut img = openImage(i);
    img.invert();
    saveImage(&img, &i, &operation);
}

fn binary_threshold(i: &str, low: u8) {
    let operation = "BinaryThreshold";
    let img = openImage(i);
    let grayscale = img.grayscale();
    let (width, height) = img.dimensions();
    // create buffer to draw the image after binary threshold
    let mut buffer = image::ImageBuffer::<Rgb<u8>, Vec<u8>>::new(width, height);
    for w in 0..width {
        for h in 0..height {
            let pixel = grayscale.get_pixel(w, h);
            let rgb = pixel.to_rgb();
            let intensity = rgb.data[0];
            if low < intensity {
                buffer.get_pixel_mut(w, h).data = [255, 255, 255];
            } else {
                buffer.get_pixel_mut(w, h).data = [0, 0, 0];
            }
        }
    }
    saveBuffer(&buffer, &i, &operation);
}
